core/libcamera_app.cpp

```diff
@@ -525,6 +548,15 @@ void LibcameraApp::StartCamera()
 	if (!controls_.get(controls::Sharpness))
 		controls_.set(controls::Sharpness, options_->sharpness);
 
+	if (!controls_.get(controls::AfMode))
+		controls_.set(controls::AfMode, options_->afMode_index);
+	if (!controls_.get(controls::AfRange))
+		controls_.set(controls::AfRange, options_->afRange_index);
+	if (!controls_.get(controls::AfSpeed))
+		controls_.set(controls::AfSpeed, options_->afSpeed_index);
+	if (!controls_.get(controls::LensPosition))
+		controls_.set(controls::LensPosition, options_->lens_position);
+
```

Comment:
On thing we ought to probably do (here and above where we call controls_.set()) is to ensure we only set AF related controls when there is an actual lens present.

One possible way to test this is to do ``camera_->controls().find(&controls::AfMode)``.  This relies on Arudcam doing the right thing in their closed source modules and only registering controls::AfMode if there is a lens present.

Reply:
`AfModeAuto` has (or should have) no effect without `AfTrigger`. It seems reasonable to trigger it when the app starts, but in that case, those controls should be in OpenCamera() instead of StartCamera() to avoid repeated triggering on mode change.

I don't think we should set `LensPosition` by default, unless this control exists _and_ the user explicitly requests that it be set to an explicit value.

Reply:
I'd quite like us to define the behaviours of the various libcamera-apps in the presence of autofocus controls, and be clear what happens in the LibcameraApp methods, and what the applications must implement.

For example, libcamera-hello/libcamera-vid are pretty straightforward. They set the AfMode to the one requested, the lens position if one is explicitly given. If an AfTrigger is requested, that could do a single sweep on startup. We could do all this in StartCamera, along with all the other controls.

libcamera-still is trickier. For example, we could say that when you switch into stills mode, it completely ignores all the AF controls.  Then it's up to the application code to do things like run an AF cycle before switching to capture. If it's in CAF mode, possibly it has to pause before switching (or is that unnecessary)? Would that cover the most obvious use-cases? But this is all open to debate, you get the idea!

There will certainly be scenarios  that are not covered by a set of simple options like this. Someone will come along and say "I only want it to do an AF sweep on every 10th timelapse capture", or something (OK, that's a bit extreme, but you know what I mean!). I'd prefer not to over-complicate the libcamera-apps, and I certainly don't want to see interactive features (like "press a key to turn AF on"). There's already a solution for all that (Picamera2).

What do folks think?

Reply:
Most of these AF controls are "set and forget". I think it does makes sense to put them in either StartCamera() or OpenCamera(), for all applications to use.

The awkward ones are `AfPause`, alluded to above, and `AfTrigger`. Contrary to my previous comment, perhaps both of those should be for the application to control. We might also want to add an option for libcamera-still to wait until AF is stable (state _not_ equal to AfStateScanning) before capturing a still; but those things could be added in a later commit.

Reply:
done